### PR TITLE
Clang-Format

### PR DESCRIPTION
**Pull request by offa**
Clang-Format configuration and a related CMake target (#102). The latter is only available if `clang-format` is found.

The new target makes it easy to format all sources:  use e.g. `make clang-format`.

### To Do

- [x] Clang Format config
- [x] Review the formatting settings
- [x] Make it easy to use
- [x] Reformat code

Not everything is matching the current formatting perfectly. E.g. multiline parameters are formatted a little different, but braces, blocks, indention should be ok (… mostly).

Please let me know what you think or what has to be changed.

### Links

- https://zed0.co.uk/clang-format-configurator
- https://clang.llvm.org/docs/ClangFormatStyleOptions.html

### CODE REVIEW

**Comment by offa** [comment]
@mattgodbolt I have rebased onto `master` and added another commit which formats the source code. Please let me know, where it isn't ok.

**Comment by mattgodbolt** [comment]
Hi @offa - happy for this to merge in (sorry for the merge nightmare, ought to be quite straightforward to re-format though after dropping the source changes).

Some comments on the style:
* I've been a fan before of keeping short statements on one line. But I get that's probably my own taste: I think it's probably best to keep the settings you've had.
* The continuation indent being the same as the actual indent means function parameter spillage/constructor initialiser list sometimes "bleeds" into the indented function body. Not too worried though.

In short: I think you should merge this in. Then I'll fiddle the CLion settings to look as close as possible to this.

Sorry for the very long delay on this! Trying to catch up with my inbox.

**Comment by offa** [comment]
> Hi @offa - happy for this to merge in (sorry for the merge nightmare, ought to be quite straightforward to re-format though after dropping the source changes).

I'm going to drop the "format everything" commit, rebase onto current master and then do the formatting again – nothing to merge then :smile: .

> The continuation indent being the same as the actual indent means function parameter spillage/constructor initialiser list sometimes "bleeds" into the indented function body. Not too worried though.

Maybe we can reduce this by putting the first argument on the same line as the ctor. But I worry that this wont be perfect or much better.

I've tested some options, but nothing really useful yet.

**Comment by offa** [comment]
Maybe this can help:

> 
> ConstructorInitializerIndentWidth (unsigned)
>     The number of characters to use for indentation of constructor initializer lists.

http://releases.llvm.org/4.0.0/tools/clang/docs/ClangFormatStyleOptions.html

#### Unformatted:
```
    explicit MyHandler(Server* server) : _server(server), _currentValue(0) {
        setValue(1);
    }
```

#### Formatted (using default value):

```
    explicit MyHandler(Server* server)
        : _server(server), _currentValue(0) {
        setValue(1);
    }
```

#### Formatted (using ConstructorInitializerIndentWidth = 8):

```
    explicit MyHandler(Server* server)
            : _server(server), _currentValue(0) {
        setValue(1);
    }
```

**Comment by mattgodbolt** [comment]
That looks perfect! Merge away :)
